### PR TITLE
Fix memory leak in component metadata rule executor

### LINKED ISSUE #12843: Cached component metadata rules leak memory
After investigating a customer memory snapshot during IDE sync we realized that if a build makes use of `@CacheableRule` then we can leak classloaders between builds.

The problem seems to be that realized component metadata that we cache cross-build in memory keep a reference to the object factory, which basically leaks whole classloaders.

Waiting for feedback from the customer on a couple of potential fixes.

### PR DESCRIPTION
**Pull request by melix**
The cross-build cache for cached component metadata rules stores
instances of realized Maven/Ivy module metadata in memory. The
problem is that those instances shouldn't have any state from
the project/Gradle build attached to them (they should be good
old POJOs) but for implementation reasons, they are not. In
particular, they keep a reference to the attributes factory,
which in turn keeps a reference to the object factory, which,
itself, will leak classloaders.

As a consequence, as more builds get executed, it was _possible_
that the cache would release some entries, but for the classloaders
to be released, it would require _all_ the entries from the same
build to be released from the cache. It turns out this event is
unlikely and as a consequence, builds end up running out of
memory.

This commit fixes the problem by making sure that at the end of
a build, we clear the in-memory cache of component metadata rules.
Therefore, cross-build, we would still read the result of the
execution of rules from the binary, on disk cache, but we would
not reuse in-memory results from previous builds.

This fix has been tested and validated by a user facing this
issue. Memory usage dropped from 11GB of memory to 1GB.

Fixes #12843

### CODE REVIEW

**Comment by melix** [comment]
@bot-gradle test FRM please

**Comment by melix** [comment]
@bot-gradle test RFM please

**Comment by ljacomet** [comment]
> Memory usage dropped from 11GB of memory to 1GB.

Well, that's a nice result indeed! 🎉

**Comment by melix** [comment]
@bot-gradle test RFM please

**Comment by blindpirate** [comment]
@bot-gradle test RFM please

File: subprojects/persistent-cache/persistent-cache.gradle.kts [hunk 1]
```diff
@@ -46,3 +46,6 @@ gradlebuildJava {
     moduleType = ModuleType.CORE
 }
 
+tasks.withType<ValidatePlugins>().configureEach {
+    enabled = false
```
Comment (jjohannes) [inline subprojects/persistent-cache/persistent-cache.gradle.kts:50]: Why is this required?

Comment (melix) [reply subprojects/persistent-cache/persistent-cache.gradle.kts:50]: This is a temporary fix because our validate plugins task doesn't handle classloading properly. I changed a type from a concrete type to an interface and it doesn't like it. Once we have a new wrapper this must be removed

**Comment by melix** [comment]
> Do you think there is a performance impact as some (flaky) perf tests failed? I can't see how this should impact the performance there though...

I don't think it does. How do you see flaky tests failing? All I can see is a misconfiguration error (databaseUrl not set).

**Comment by jjohannes** [comment]
> I don't think it does. How do you see flaky tests failing? All I can see is a misconfiguration error (databaseUrl not set).

Sorry. Then I looked at the wrong build (TC UI is driving me crazy). Then it's all good.

**Comment by melix** [comment]
I'm going to build a new wrapper soon so that I can remove the hack.
